beacon/blockchain/payload_test.go

```diff
@@ -0,0 +1,450 @@
+)
+
+// When we reject a block and we have optimistic payload building enabled
+// we must make sure that a few beacon state quantities are duly pre-processed
+// before building the block.
+func TestOptimisticBlockBuildingRejectedBlockStateChecks(t *testing.T) {
+	t.Parallel()
+
+	optimisticPayloadBuilds := true // key to this test
+	cs, err := spec.MainnetChainSpec()
+	require.NoError(t, err)
+
+	sp, st, depStore, ctx, _, eng := statetransition.SetupTestState(t, cs)
+
+	logger := log.NewNopLogger()
+	ts := metrics.NewNoOpTelemetrySink()
+	sb := mocks.NewStorageBackend(t)
+	sb.EXPECT().StateFromContext(mock.Anything).Return(st)
+	sb.EXPECT().DepositStore().RunAndReturn(func() *depositstore.KVStore { return depStore })
+
+	b := bmocks.NewLocalBuilder(t)
+	b.EXPECT().Enabled().Return(optimisticPayloadBuilds)
+
+	chain := blockchain.NewService(
+		sb,
+		nil, // blockchain.BlobProcessor unused in this test
+		nil, // deposit.Contract unused in this test
+		logger,
+		cs,
+		eng,
+		b,
+		sp,
+		ts,
+		optimisticPayloadBuilds,
+	)
+	// Note: test avoid calling chain.Start since it only starts the deposits
+	// goroutine which is not really relevant for this test
+
+	// Before processing any block it is mandatory to handle genesis
+	// TODO: I had to manually align default genesis and cs specs
+	// Check if this is correct/necessary
+	genesisData := ctypes.DefaultGenesis(cs.GenesisForkVersion())
+
+	genesisData.ExecutionPayloadHeader.Timestamp = math.U64(cs.GenesisTime())
+	genesisData.Deposits = []*ctypes.Deposit{
+		{
+			Pubkey: [48]byte{0x01},
+			Amount: cs.MaxEffectiveBalance(),
+			Credentials: ctypes.NewCredentialsFromExecutionAddress(
+				common.ExecutionAddress{0x01},
+			),
+			Index: uint64(0),
+		},
+	}
+	genBytes, err := json.Marshal(genesisData)
+	require.NoError(t, err)
+	_, err = chain.ProcessGenesisData(ctx.ConsensusCtx(), genBytes)
+	require.NoError(t, err)
+
+	// Finally create a block that will be rejected and
+	// verify the state on top of which is next payload built
+	var (
+		consensusTime   = math.U64(time.Now().Unix())
+		proposerAddress = []byte{'d', 'u', 'm', 'm', 'y'} // this will err on purpose
+	)
+
+	// Since this is the first block called post genesis
+	// forceSyncUponProcess will be called.
+	dummyPayloadID := &engineprimitives.PayloadID{1, 2, 3}
+	eng.EXPECT().NotifyForkchoiceUpdate(mock.Anything, mock.Anything).Return(dummyPayloadID, nil)
+
+	// we set just enough data in invalid block to let it pass
+	// the first validations in chain before state processor is invoked
+	invalidBlk := &ctypes.BeaconBlock{
+		Slot: 1, // first block after genesis
+		Body: &ctypes.BeaconBlockBody{
+			ExecutionPayload: &ctypes.ExecutionPayload{
+				Timestamp: math.U64(cs.GenesisTime() + 1),
+			},
+		},
+	}
+
+	// register async call to block building
+	var wg sync.WaitGroup          // useful to make test wait on async checks
+	var ch = make(chan struct{})   // useful to serialize build block goroutine and avoid data races
```

Comment:
without this, UT seems to be having data races, but this is concerning.
The whole PR increases access to the state during the block building, but state is already access currently.
Do we already have a data race that this refactoring highlights?
Or does the data race happens in this refactoring?
Or is the data race detector wrong (like this can only happen in tests, not in prod code)?